### PR TITLE
order support

### LINKED ISSUE #1735: Add support for `Order`
Related to https://github.com/hibernate/hibernate-reactive/issues/1706

See the new class in ORM: [org.hibernate.query.Order](https://github.com/hibernate/hibernate-orm/blob/main/hibernate-core/src/main/java/org/hibernate/query/Order.java)

### PR DESCRIPTION
**Pull request by blafond**
fixes #1735
 - implements and tests reactive session's `setOrder(...)`

### CODE REVIEW

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+		return List.of(
+				Book.class
+		);
+	}
+
+	static boolean skipDBs() {
+		// MySQL & MariaDB will fail because the simple 'String isbn' element will prevent book from being created correctly
+		// errorMessage=Table 'hreact.Book' doesn't exist, errorCode=1146
+		// NOTE: ORM's OrderTest will also fail with these DB's
+		return DatabaseConfiguration.dbType() == MYSQL ||
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java:51]: If I run the test with MySQL it works fine. There was probably something wrong on your machine.

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+	private CompletionStage<Void> populateDB() {
+		return getSessionFactory()
+				.withTransaction( session -> session.persist( book1, book2 ) );
+	}
+
+	private Uni<Void> populateDBMunity() {
+		return getMutinySessionFactory()
+				.withSession( session -> session.persistAll( book1, book2 )
+						.chain( session::flush ) );
+	}
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java:64]: There's no reason to have two separate methods for adding entries in the database.
We are testing the order, so you can just inert the elements once ([like we do everywhere else](https://github.com/hibernate/hibernate-reactive/blob/11950394f8cf943a28e00d2cfd99c21cc58e31b4/hibernate-reactive-core/src/test/java/org/hibernate/reactive/HQLQueryTest.java#L42)).

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+	@Test
+	public void testOrder(VertxTestContext context) {
+		test(
+				context,
+				populateDB().thenCompose( v -> getSessionFactory()
+						.withTransaction( (session, tx) -> session
+								.createSelectionQuery( "select title from Book", String.class )
+								.getResultList()
+								.thenAccept( list -> assertTrue( list.size() == 2 ) )
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java:74]: Why did you add this block?
The purpose of the test class is to check the `setOrder` method, and you are not using it here.

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+											.findEntityType( Book.class );
+									SingularAttribute<? super Book, ?> title = bookType.findSingularAttribute( "title" );
+									return session
+											.createSelectionQuery( "from Book", Book.class )
+											.setOrder( Order.asc( title ) )
+											.getResultList();
+								} ).thenAccept( books -> {
+									assertEquals( "Hibernate in Action", books.get( 0 ).title );
+									assertEquals( "Java Persistence with Hibernate", books.get( 1 ).title );
+								} )
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java:87]: with AssertJ, you can replace this assertion with:
```
assertThat( books ).containsExactly(book1, book2);
```

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+								.thenCompose( vv -> {
+									MappingMetamodelImpl metamodel = (MappingMetamodelImpl) getSessionFactory().getMetamodel();
+									EntityDomainType<Book> bookType = metamodel.getJpaMetamodel()
+											.findEntityType( Book.class );
+									SingularAttribute<? super Book, ?> title = bookType.findSingularAttribute( "title" );
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java:92]: There's no reason to repeat these three lines everywhere: you can declare them once at the begging of the test

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java (hunk 4 above)
Comment (DavideD) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java:87]: There are other places where you can do the same

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+								} ).thenAccept( books -> {
+									assertEquals( "Hibernate in Action", books.get( 0 ).title );
+									assertEquals( "Java Persistence with Hibernate", books.get( 1 ).title );
+								} )
+						) )
+		);
+	}
+
+	@Test
+	public void testAscendingDescendingBySelectElement(VertxTestContext context) {
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java:132]: Maybe it's better to test this with Mutiny as well

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+									SingularAttribute<? super Book, ?> isbn = bookType.findSingularAttribute( "isbn" );
+									return session
+											.createSelectionQuery( "from Book", Book.class )
+											.setOrder( Order.desc( isbn ) )
+											.getResultList();
+								} ).invoke( books -> {
+									assertEquals( "Hibernate in Action", books.get( 0 ).title );
+									assertEquals( "Java Persistence with Hibernate", books.get( 1 ).title );
+								} )
+				) )
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java:224]: If you create a single test method for each block, the whole class becomes much more readable and easier to understand (same is true for the other tests).

**Review by DavideD — CHANGES_REQUESTED** [review]
Looks fine overall.

Maybe I've missed it, but I don't think there's a test for `setOrder(List<Order>)`

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java (hunk 1 above)
Comment (blafond) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java:51]: FYI... yes.. clearing my docker containers and rebooting... all DB's work

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java (hunk 6 above)
Comment (DavideD) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java:132]: On second thought, nevermind. Ignore that

**Comment by blafond** [comment]
I'll have an update in the morning to the test.  Took your suggestions and it's looking cleaner and added a couple of tests, including the List<> case.

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java (hunk 3 above)
Comment (blafond) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java]: removed

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java (hunk 2 above)
Comment (blafond) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java:53]: Can we mix Mutiny and Stage methods within the `test(...)` method?

Comment (DavideD) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java:53]: No, but you can use the `@BeforeEach` to run the inserts before running the actual tests. Like in the test class I linked in the previous comment

Comment (blafond) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderTest.java:46]: done

**Comment by DavideD** [comment]
Looks good.
 @blafond, could you check if [these changes](https://github.com/DavideD/hibernate-reactive/commit/32ce11e1ad138c9ba22004c94a28c55a71a606d5) make sense to you?

**Comment by DavideD** [comment]
Superseded by https://github.com/hibernate/hibernate-reactive/pull/1815
